From 4930927c62f705b8912c5320267b6016817966d8 Mon Sep 17 00:00:00 2001
From: Aaradhana Sahu <aaradhana.sahu@oss.qualcomm.com>
Date: Fri, 19 Sep 2025 15:27:09 +0530
Subject: [PATCH] UPSTREAM:CP: wifi: ath12k: Reduce power consumption when
 radio is inactive

Currently, when the wifi down command is executed, power consumption
remains high even if no radio is active. To address this, the driver
sends a WMI_PDEV_SUSPEND_CMDID command to the firmware whenever the
radio becomes inactive.

When the radio is restarted with the wifi command, and the pdev is in a
suspended state, the driver issues a WMI_PDEV_RESUME_CMDID to the firmware.

Additionally, move ath12k_mac_pdev_suspend() from ath12k_mac_vdev_delete()
to ath12k_mac_stop(), as it is invoked for all inactive radios during
ERP and wifi down operations. Also, relocate ath12_info log inside the
conditional block in ath12k_core_power_up() to improve code readability.

Patch-dependency:
0441-03-UPSTREAM-CP-wifi-ath12k-Enable-pdev-suspend-and-resu.patch
Patch-work: None

Signed-off-by: Aaradhana Sahu <aaradhana.sahu@oss.qualcomm.com>
---
 drivers/net/wireless/ath/ath12k/core.c |  2 +-
 drivers/net/wireless/ath/ath12k/mac.c  | 63 ++++++++++++++++++--------
 drivers/net/wireless/ath/ath12k/mac.h  |  1 +
 3 files changed, 47 insertions(+), 19 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/core.c b/drivers/net/wireless/ath/ath12k/core.c
index 230f053..975c02a 100644
--- a/drivers/net/wireless/ath/ath12k/core.c
+++ b/drivers/net/wireless/ath/ath12k/core.c
@@ -974,11 +974,11 @@ int ath12k_core_power_up(struct ath12k_hw_group *ag)
 
 	for (i = 0; i < ag->num_probed; i++) {
 		ab =  ag->ab[i];
-		ath12k_info(ab, "Q6 power up is started\n");
 		if (ab->pm_suspend) {
 			ath12k_hif_power_up(ab);
 			ab->pm_suspend = false;
 			ab->powerup_triggered = true;
+			ath12k_info(ab, "Q6 power up is started\n");
 			reinit_completion(&ab->power_up);
 		}
 	}
diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index af6bd42..736ee6a 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -14936,6 +14936,16 @@ int ath12k_mac_start(struct ath12k *ar)
 
 	rcu_assign_pointer(ab->pdevs_active[ar->pdev_idx],
 			   &ab->pdevs[ar->pdev_idx]);
+
+	if (ar->pdev_suspend && (!ath12k_check_erp_power_down(ab->ag) ||
+	    !ab->powerup_triggered)) {
+		ret = ath12k_mac_pdev_resume(ar);
+		if (ret) {
+			ath12k_err(ab, "pdev resume command is failed: %d\n", ret);
+			goto err;
+		}
+	}
+
 	return 0;
 err:
 
@@ -15136,6 +15146,12 @@ void ath12k_mac_stop(struct ath12k *ar)
 	if (!list_empty(&dp->neighbor_peers))
 		ath12k_debugfs_nrp_cleanup_all(ar);
 
+	if (!ar->allocated_vdev_map && !ar->pdev_suspend) {
+		ret = ath12k_mac_pdev_suspend(ar);
+		if (ret)
+			ath12k_warn(ar->ab, "pdev suspend command is failed %d\n", ret);
+	}
+
 	rcu_assign_pointer(ar->ab->pdevs_active[ar->pdev_idx], NULL);
 
 	synchronize_rcu();
@@ -15510,6 +15526,30 @@ void ath12k_mac_11d_scan_stop_all(struct ath12k_base *ab)
 	}
 }
 
+int ath12k_mac_pdev_resume(struct ath12k *ar)
+{
+	unsigned long time_left;
+	int ret;
+
+	reinit_completion(&ar->pdev_resume);
+	ret = ath12k_wmi_pdev_resume(ar, ar->pdev->pdev_id);
+
+	if (ret) {
+		ath12k_err(ar->ab, "failed to send wmi resume command %d\n", ret);
+		return ret;
+	}
+
+	time_left = wait_for_completion_timeout(&ar->pdev_resume,
+						ATH12K_PDEV_RESUME_TIMEOUT);
+	if (!time_left) {
+		ath12k_err(ar->ab, "timeout in receiving pdev resume response %d\n",
+			   ar->pdev->pdev_id);
+		return -ETIMEDOUT;
+	}
+
+	return 0;
+}
+
 int ath12k_mac_vdev_create(struct ath12k *ar, struct ath12k_link_vif *arvif,
 			   bool is_bridge_vdev)
 {
@@ -15532,7 +15572,6 @@ int ath12k_mac_vdev_create(struct ath12k *ar, struct ath12k_link_vif *arvif,
 	u8 mask[ETH_ALEN] = {0xFF, 0xFF, 0xFF, 0x00, 0x00, 0x00};
 	int txpower = NL80211_TX_POWER_AUTOMATIC;
 	u8 map_id;
-	unsigned long time_left;
 	u32 rep_ul_resp;
 
 	lockdep_assert_wiphy(hw->wiphy);
@@ -15544,21 +15583,12 @@ int ath12k_mac_vdev_create(struct ath12k *ar, struct ath12k_link_vif *arvif,
 		return -EINVAL;
 
 	if (ar->pdev_suspend) {
-		reinit_completion(&ar->pdev_resume);
-		ret = ath12k_wmi_pdev_resume(ar, ar->pdev->pdev_id);
-
+		ret = ath12k_mac_pdev_resume(ar);
 		if (ret) {
-			ath12k_err(ar->ab, "failed to send wmi resume command %d\n", ret);
+			ath12k_err(ab,
+				   "vdev could not be created because the pdev failed to resume\n");
 			return ret;
 		}
-
-		time_left = wait_for_completion_timeout(&ar->pdev_resume,
-							ATH12K_PDEV_RESUME_TIMEOUT);
-		if (!time_left) {
-			ath12k_err(ar->ab, "Timeout in receiving pdev resume response: %d\n",
-				   ar->pdev->pdev_id);
-			return -ETIMEDOUT;
-		}
 	}
 
 	/* If no link is active and scan vdev is requested
@@ -16356,7 +16386,8 @@ int ath12k_mac_pdev_suspend(struct ath12k *ar)
 	time_left = wait_for_completion_timeout(&ar->suspend,
 						ATH12K_PDEV_SUSPEND_TIMEOUT);
 	if (!time_left) {
-		ath12k_err(ar->ab, "Timeout in receiving pdev suspend response: %d\n", ar->pdev->pdev_id);
+		ath12k_err(ar->ab, "timeout in receiving pdev suspend response %d\n",
+			   ar->pdev->pdev_id);
 		ret = -ETIMEDOUT;
 		goto exit;
 	}
@@ -16452,10 +16483,6 @@ err_vdev_del:
 
 	if (!ar->allocated_vdev_map && !arvif->is_scan_vif) {
 		if (ath12k_erp_get_sm_state() == ATH12K_ERP_ENTER_COMPLETE) {
-			ret = ath12k_mac_pdev_suspend(ar);
-			if (ret)
-				ath12k_warn(ab, "Pdev suspend command is failed %d\n", ret);
-
 			if (ath12k_mac_validate_active_radio_count(ar->ah))
 				ath12k_core_cleanup_power_down_q6(ar->ah);
 		}
diff --git a/drivers/net/wireless/ath/ath12k/mac.h b/drivers/net/wireless/ath/ath12k/mac.h
index 66c4f84..9dfbfa7 100644
--- a/drivers/net/wireless/ath/ath12k/mac.h
+++ b/drivers/net/wireless/ath/ath12k/mac.h
@@ -479,6 +479,7 @@ bool ath12k_mac_validate_active_radio_count(struct ath12k_hw *ah);
 int ath12k_mac_pdev_suspend(struct ath12k *ar);
 int ath12k_mac_set_eht_txbf_conf(struct ath12k_link_vif *arvif);
 int ath12k_mac_set_he_txbf_conf(struct ath12k_link_vif *arvif);
+int ath12k_mac_pdev_resume(struct ath12k *ar);
 
 /* In the bitmap 0 indicates no puncturing and 1 indicated that sub channel is
  * punctured
-- 
2.34.1

